overlord/snapstate/handlers.go

```diff
@@ -1150,59 +1150,190 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 		return copyDataErr
 	}
 
-	// the migration hasn't been done - do it now
-	if opts.UseHidden && !opts.MigratedToHidden {
-		if err = m.backend.HideSnapData(snapsup.InstanceName()); err != nil {
+	var oldBase string
+	if oldInfo != nil {
+		oldBase = oldInfo.Base
+	}
 
-			// undo the migration. In contrast to copy data for which the new revision
-			// directory can just be discarded, the snap data migration introduces
-			// a change that affects all revisions of the snap and thus needs to be reverted
-			if err := m.backend.UndoHideSnapData(snapsup.InstanceName()); err != nil {
-				src := filepath.Join("~", dirs.HiddenSnapDataHomeDir, snapsup.InstanceName())
-				dst := filepath.Join("~", dirs.UserHomeSnapDir, snapsup.InstanceName())
+	action := triggeredMigration(oldBase, newInfo.Base, opts)
+	var undo func()
 
-				st.Lock()
-				st.Warnf("cannot undo snap dir hiding (move all user's %s to %s): %v",
-					src, dst, err)
-				st.Unlock()
-			}
+	switch action {
+	case full:
+		undo, err = m.doFullMigration(t, st, snapsup, newInfo)
+	case hidden:
+		undo, err = m.doHiddenMigration(t, st, snapsup, newInfo)
+	case home:
+		undo, err = m.doHomeMigration(t, st, snapsup, newInfo.Revision)
+	case revertHidden:
+		undo, err = m.revertHiddenMigration(t, st, snapsup)
+	default:
+		// we don't support reverting the ~/Snap migration during a refresh
+		err = nil
+	}
 
-			return err
+	// if task fails, undo any migration actions taken
+	defer func() {
+		if err != nil && undo != nil {
+			undo()
 		}
+	}()
 
-		snapsup.MigratedHidden = true
-		st.Lock()
-		err := SetTaskSnapSetup(t, snapsup)
-		st.Unlock()
-		if err != nil {
-			return fmt.Errorf("cannot set migration status to done: %w", err)
+	return err
+}
+
+type migration string
+
+const (
+	// none states that no action should be taken
+	none migration = "none"
+	// full migrates ~/snap to ~/.snap and the new home to ~/Snap
+	full migration = "full"
+	// hidden migrates ~/snap to ~/.snap
+	hidden migration = "hidden"
+	// home migrates the new home to ~/Snap
+	home migration = "home"
+	// revertHidden undoes the hidden migration (i.e., moves ~/.snap to ~/snap)
+	revertHidden migration = "revertHidden"
```

Comment:
Will we have a `revertFull` as well? Just wondering, if instead of these strings we should have a bitfield with flags, since it seems to me that some of these actions can have some overlapping behaviour. (this is a question, not a request for change :-) )

Reply:
Yes, there will be a revertFull and even a revertHome so I agree that those would be well expressed as a bitfield. However, the remaining values can't be combined so we'd have to add more code to "forbid" that. One thing I maybe should've made a comment about is that these values are strings because it makes test failures easier to understand. They were iotas initially and the errors looked like "expected 0 got 5".